partition_balancer: unify disk usage limits into a single threshold

bharathv opened the pull request:
Previously, the partition balancer used two disk usage thresholds:

1. `partition_autobalancing_max_disk_usage_percent` (soft limit): triggers moves *out* of overloaded brokers.

2. `storage_space_alert_free_threshold_percent` (hard limit): prevents moves *into* brokers if usage would exceed this limit after a move.

The configuration that is used for hard limit enforcement is intended for storage alerts, not balancing. When set too conservatively, it can stall the balancer entirely (unintentionally)

This change removes the distinction between soft and hard limits, replacing them with a single, dedicated threshold used for both move-out and move-in decisions. This simplifies configuration and avoids unintended coupling with alerting behavior.

An alternative considered was introducing a new hard-limit setting specific to balancing. This was rejected to avoid increasing configuration sprawl. In rare cases (e.g., usage >80%), administrators can still temporarily raise the threshold to enable recovery.

This is effectively a revert of commit 5f65dd0eeb8a5d05

Fixes: CORE-8572

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x

## Release Notes
### Improvements

* Simplified partition balancer disk usage limits by consolidating soft and hard thresholds into a single configurable setting. `storage_space_alert_free_threshold_percent` is no longer used for partition balancer decisions.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
partition_balancer: unify disk usage limits into a single threshold

Previously, the partition balancer used two disk usage thresholds:

1. `partition_autobalancing_max_disk_usage_percent` (soft limit): triggers
   moves *out* of overloaded brokers.

2. `storage_space_alert_free_threshold_percent` (hard limit): prevents
   moves *into* brokers if usage would exceed this limit after a move.

The config backing hard limit is intended for storage alerts, not balancing.
When set too conservatively, it can stall the balancer entirely
(unintentionally)

This change removes the distinction between soft and hard limits,
replacing them with a single, dedicated threshold used for both move-out
and move-in decisions. This simplifies configuration and avoids
unintended coupling with alerting behavior.

An alternative considered was introducing a new hard-limit setting
specific to balancing. This was rejected to avoid increasing
configuration sprawl. In rare cases (e.g., usage >80%), administrators
can still temporarily raise the threshold to enable recovery.

This is effectively a revert of commit 5f65dd0eeb8a5d05

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73203-->
<!--79d9be33-7fe6-443a-92a9-092cbef202f9-->
<!--title-->
## Retry command for [Build#73203](https://buildkite.com/redpanda/redpanda/builds/73203)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_full_nodes
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c83fc88d-0814-4318-9833-4b26f3f8660f-->
<!--TEST RESULTS-->
<!--b331c691-71b1-4bd3-9d5d-656b3aac2d33-->
<!--TEST RESULTS-->
<!--54523bc5-e4e3-474a-a9cc-81a0cf0e2ab8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73203">build#73203</a></summary>


| test_class            | test_method     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                        |
|:----------------------|:----------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest | test_full_nodes | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73203#0199982a-05ef-449a-8378-c2772116b110 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_full_nodes |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73213">build#73213</a></summary>


| test_class                           | test_method                | test_arguments                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:-------------------------------------|:---------------------------|:----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true}                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73213#019998ed-183e-447d-978a-160efbcac3b0 | FLAKY         | 20/21    | upstream reliability is '96.98795180722891'. current run reliability is '95.23809523809523'. drift is 1.74986 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| PartitionBalancerTest                | test_full_nodes            | null                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73213#019998ed-183d-4149-81ee-2e6087061265 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_full_nodes                           |
| TieredStorageTest                    | test_tiered_storage        | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73213#019998ed-183f-4e00-aa9f-e6c68d68d23c | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73237">build#73237</a></summary>


| test_class                          | test_method          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:------------------------------------|:---------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest               | test_full_nodes      | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73237#01999b3b-eae6-4a85-9a22-d8238d187f71 | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_full_nodes              |
| RandomNodeOperationsTest            | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73237#01999b3b-eadf-44e1-a4b7-0151763b33f5 | FLAKY         | 19/21    | upstream reliability is '99.16666666666667'. current run reliability is '90.47619047619048'. drift is 8.69048 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations      |
| WriteCachingFailureInjectionE2ETest | test_crash_all       | {"use_transactions": false}                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/73237#01999b3b-eae9-4aad-8ec1-4a59a3097ca1 | FLAKY         | 15/21    | upstream reliability is '87.74373259052925'. current run reliability is '71.42857142857143'. drift is 16.31516 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73213-->
<!--dc4ecae5-c6c1-45f9-9076-39fd756ba951-->
<!--title-->
## Retry command for [Build#73213](https://buildkite.com/redpanda/redpanda/builds/73213)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_full_nodes
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73237-->
<!--d4c667ad-21b6-42f7-ada4-88fab27f69ed-->
<!--title-->
## Retry command for [Build#73237](https://buildkite.com/redpanda/redpanda/builds/73237)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_full_nodes
```
<!--/body-->

bharathv committed:
partition_balancer/tests: remove full_nodes_test

Without a soft/hard distinction the balancer will stall and not move
beyond 80%, so the test no longer makes sense.

bharathv commented:
/dt

joe-redpanda reviewed (approved):
LGTM

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27798-v24.3.x-671 remotes/upstream/v24.3.x
git cherry-pick -x b98648835d 2fb5aa20fd
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18196896164).

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27798-v25.1.x-369 remotes/upstream/v25.1.x
git cherry-pick -x b98648835d 2fb5aa20fd
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18196898339).
